sklearn/feature_extraction/tests/test_text.py

```diff
@@ -1196,3 +1197,51 @@ def build_preprocessor(self):
                                             .findall(doc),
                     stop_words=['and'])
     assert _check_stop_words_consistency(vec) is True
+
+
+@pytest.mark.parametrize('Estimator',
+                         [CountVectorizer, TfidfVectorizer, HashingVectorizer])
+def test_callable_analyzer_vs_file_input(Estimator):
+    data = ['this is text, not file or filename']
+
+    with pytest.raises(FileNotFoundError):
+        Estimator(analyzer=lambda x: x.split(),
+                  input='filename').fit_transform(data)
+
+    with pytest.raises(AttributeError,
+                       match="'str' object has no attribute 'read'"):
+        Estimator(analyzer=lambda x: x.split(),
+                  input='file').fit_transform(data)
+
+    # check if the ChangedBehaviorWarning is raised if the given analyzer
+    # expects a file or a file name.
+    def analyzer1(doc):
+        open(doc, 'r')
+
+    def analyzer2(doc):
+        doc.read()
+
+    for analyzer in [analyzer1, analyzer2]:
+        for input_type in ['file', 'filename']:
+            try:
+                with pytest.warns(ChangedBehaviorWarning,
+                                  match="Since v0.21, vectorizer"):
+                    Estimator(analyzer=analyzer,
+                              input=input_type).fit_transform(data)
+            except Exception:
+                pass
+
+    # check if a custom exception from the analyzer is shown to the user
+    def analyzer3(doc):
+        raise Exception("testing")
+
+    import tempfile
+    with tempfile.NamedTemporaryFile('w', delete=False) as f:
+        f.write("sample content\n")
+        fname = f.name
+
+    with pytest.raises(Exception, match="testing"):
+        Estimator(analyzer=analyzer3, input='filename').fit_transform([fname])
+
+    import os
+    os.remove(fname)
```

Comment:
Better to use pytest `tmpdir` fixture

Reply:
done

Reply:
although I personally find it kinda magic-like, and probably confusing for somebody who sees the code later.

Reply:
I agree that too many fixtures get somewhat magical, but for temporary dirs it is quite useful IMO and we use it constantly through the code.